File: src/items/NormalItem.py
from src.items.Item import *
from src.items.Interface import *

'''
This class will adquire the propierties about the Item class, which give it the atributes
and the interface class, that making a comparation with java, we will need a constraint 
to keep our software under the desing per contract => We need updateables items to continue.
'''
class NormalItem(Item, Interface):

    # Constructor
    def __init__(self, name, sell_in, quality):

        Item.__init__(self, name, sell_in, quality)


    # Getters/Setters
    def getName(self):

        return self.name

    def getSellIn(self):

        return self.sell_in

    def getQuality(self):

        return self.quality

    def setName(self, new_name):

        self.name = new_name

    def setSellIn(self):

        self.sell_in = self.sell_in + -1

    # By domain, the quality has the following behaviour (check ProblemSpecificationRequirements.txt)
    def setQuality(self, quality_increment):

        if self.quality < 50:

            self.quality = 50

        elif self.quality < 0:

            self.quality = 0

        else:

            self.quality += quality_increment

        assert 0 < self.quality <= 50 # Invariant, the quality must be always under those values

File: java/src/domain/NormalItem.java
package domain;


public class NormalItem implements Updateable {
    
    private final Item item;

    // Constructor to the method
    public NormalItem(String name, int sellIn, int quality){

        this.item = new Item(name, sellIn, quality);
    }

    // Override the method toString
    @Override
    public String toString(){
        return item.toString();
    }

    // Getters and Setters
    public Item getItem(){

        return this.item;
    }

    public String getName(){
        
        return item.name;
    }

    public int getSellIn(){

        return item.sellIn;
    }

    public int getQuality(){

        return item.quality;
    }

    public void setName(String newName){

        item.name = newName;
    }

    public void setSellIn(){

        item.sellIn = item.sellIn - 1;
    }

    public void setQuality(int value){

        item.quality = value;
    }

    // Calculate the quality if the value is between 0 and 50. Keeping encapsulation.
    public void calculateQuality(int value){
        
        if (getQuality() + value > 50){

            setQuality(50);
        }

        else if (getQuality() + value >= 0){

            setQuality(getQuality()+value);
        }

        else {

            setQuality(0);
        }

    }

    // Override the method updateQuality to this class.
    @Override
    public void updateQuality(){

        if (getSellIn()>0){
            calculateQuality(-1);
        }

        else if (getSellIn()<0){
            calculateQuality(-2);
        }

        this.setSellIn();
    }

}
